### ISSUE TITLE
How to add a lock timeout to migrations?

### ISSUE THREAD

**Issue by samphilipd** [elixir-ecto/ecto#2801]
We ran into a scenario recently where we had downtime for the following reason:

1. We had a long running transaction open
2. We had a migration that attempted to take an ACCESS EXCLUSIVE lock (happens for all `ALTER TABLE` commands)

This caused all queries to be queued behind the migration, which was waiting to get a lock.

We'd like to automatically set a lock timeout on migrations to avoid this problem in future. The command is:

`SET lock_timeout TO '5s'`

We could add this manually inside every migration we create but that is tedious and prone to people forgetting. Is there a way to use ecto to automatically execute this SQL at the start of every DDL transaction?

**Comment by josevalim** [comment]
You can use `after_connect` in the repository and set this based on an environment variable.

    after_connect: {MyApp.Repo, :after_connect, 1}

and then:

    def after_connect(conn) do
      if System.get_env("REPO_LOCK_TIMEOUT") do
        Postgrex.query!(conn, ...)
      end
      conn
    end

Alternatively we can set a timeout on Elixir side of things... although I am not sure if it helps.

**Comment by samphilipd** [comment]
Is `after_connect` executed in the context of the DDL transaction?

**Comment by josevalim** [comment]
No, it is invoked whenever the connection is established. Does it have to
be set per migration?
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by samphilipd** [comment]
@josevalim yes, it must be executed inside the transaction where the migration statements are fired.
